New config file feature

mprenditore opened the pull request:
I've worked on that last year and I found it useful.

I've added the possibility to have a config file to easily configure it without messing with the script it self.
That make the script even more portable and easy to manage via CMS like Ansible, Salt and so on.
I've also improved a little bit the error handling and the debug for a better troubleshoting.

Maybe it will be good for someone else too :)

Cheers!

mprenditore committed:
Added config file and added some cache file checks

mprenditore committed:
Merge remote-tracking branch 'upstream/master'

mprenditore committed:
changed default .sock file

mprenditore committed:
comments cleanup

mprenditore committed:
changed hrsp_Yxx  value type to numeri(Unsigned)

anapsix commented on haproxy_discovery.sh:
```diff
@@ -40,7 +57,7 @@ case $1 in
 		for backend in $(get_stats | grep BACKEND | cut -d, -f1 | uniq); do
-			for server in $(get_stats | grep "^${backend}," | grep -v BACKEND | cut -d, -f2); do
+			for server in $(get_stats | grep "${backend}," | grep -v BACKEND | cut -d, -f2); do
 				serverlist="$serverlist,\n"'\t\t{\n\t\t\t"{#BACKEND_NAME}":"'$backend'",\n\t\t\t"{#SERVER_NAME}":"'$server'"}'
```
wouldn't it help to be more precise here by using `^` from beginning of the line?

mprenditore replied:
indeed, I don't know if it's a typo of if I did it on purpose, but you're right, without `^` can cause bad behavior.
Thanks

mprenditore committed:
Added "^" to the SERVER discovery

anapsix commented on haproxy_discovery.sh:
```diff
@@ -11,7 +11,24 @@
+    else
+        echo "`echo -e "$CONF_MAP" | grep $1=  | xargs -d "=" | grep $1 |awk '{print $2}'`"
+    fi
+}
```
I like the idea of conf file.. 
Though, wouldn't be easier to just source the file?
What's the advantage of using `get_conf` over sourcing the file?
I mean, I see how it uses `CONF_MAP` to lookup a conf_val if it's not in config file, but setting some defaults and then overriding them by sourcing the file is way simpler. And you don't have to SHELL out  multiple times every time you need to get a variable value set via config.
```
ha_sock=/var/run/haproxy/info.sock
debug=0
discovery_log_file=/var/tmp/haproxy_discovery.log

if [[ -r $CONF_FILE ]]; then
  source ${CONF_FILE}
fi
```

mprenditore replied:
That's a good idea, I'll work on it tomorrow +1

mprenditore replied:
changed on both scripts to use ENV vars

anapsix commented:
Thanks for your contribution @mprenditore.. and apologies for much overdue response.
I've added few comment, particularly I think it would be great to simplify config file usage.

mprenditore commented:
Thank you @anapsix for the feedback. I'll try to modify it by tomorrow :)

anapsix commented on haproxy_stats.sh:
```diff
@@ -18,31 +18,71 @@ pxname="$1"
+    debug $2
+    write_log
+    exit $1
 }
```
the `write_log` logic is a bit too complicated.. let's try to simplify it
```
debug() {
  [[ "${DEBUG}" -eq 1 ]] || true  # return immediately if debug is disabled
  if [[ "${DEBUG_ONLY_LOG}" -eq 1 ]]; then
     "DEBUG: $@" >> ${LOG_FILE}
     return
   else
     echo >&2 "DEBUG: $@"
   fi
}

fail() {
  local _exit_code=${1:-1}
  shift 1
  if [[ -n "$1" ]]; then
    if [[ "${DEBUG}" -eq 0 ]]; then
      echo >&2 "$@"
    else
      debug "$@"
    fi
  fi
  exit $_exit_code
}
```
would this do what you are looking for?

anapsix replied:
if you'd like to prepend your debug log file with a date, it can be added separately when script starts
```
debug "######__________ $(date) ____________"
```

mprenditore replied:
I like the removing of the `write_log` and the new `fail` implementation but regarding the `debug` I think the following should do a better job:
```
debug() {
    [[ "${DEBUG}" -eq 1 ]] || return  # return immediately if debug is disabled
    echo "DEBUG: $@" >> ${LOG_FILE}
    [[ ${DEBUG_ONLY_LOG} -ne 1 ]] || return  # return if it should only debug to log
    echo >&2 "DEBUG: $@"
}
```

anapsix replied:
agreed, it's cleaner that way

~though I think you've meant to have `-eq` when checking `$DEBUG_ONLY_LOG` value.. as if DEBUG_ONLY_LOG is enabled, there is no reason to write to stderr~

nevermind)

mprenditore replied:
exactly what I meant, but if you use the `-eq` you have to change the `||` to `&&` or else it will write on stderr when `DEBUG_ONLY_LOG` is enabled.
```
debug() {
    [[ "${DEBUG}" -eq 1 ]] || return  # return immediately if debug is disabled
    echo "DEBUG: $@" >> ${LOG_FILE}
    [[ ${DEBUG_ONLY_LOG} -eq 1 ]] && return  # return if it should only debug to log
    echo >&2 "DEBUG: $@"
}
```

anapsix replied:
misread my own logic.. nevermind )

anapsix commented on haproxy_stats.sh:
```diff
@@ -61,28 +100,26 @@ fi
+        fail 126 "ERROR: haproxy socket is not writable"
+    fi
+elif [ ! -r "$CACHE_STATS_FILEPATH" ]; then
+    fail 126 'ERROR: cannot read stats cache file'
```
why not put the fail condition into else or the "if" you've started in line 104?

mprenditore replied:
To which `fail` are you referring to?

I didn't changed the logic of that nested `if` I've just converted the `echo`&`exit` into `fail` and added the caching file removal in case the file exists but it's empty (that happened to me a lot while using the original implementation).

anapsix replied:
ah right.. the idea was if file exists, but not readable..

anapsix commented on haproxy_stats.sh:
```diff
@@ -175,72 +212,109 @@ query_stats() {
+      CFILE=$CACHE_INFO_FILEPATH
+      CTIME=$CACHE_INFO_EXPIRATION
+  fi
+  find ${CFILE} -mmin +${CTIME} -delete >/dev/null 2>&1
```
I've used `stat` command cause it was providing more precise modification time, than `find` can.
seconds vs minutes..
also, it appears you dislike `flock`
could you explain rationale behind your approach, removing `flock` and switching to `find`?

anapsix commented on haproxy_stats.sh:
```diff
@@ -175,72 +212,109 @@ query_stats() {
-  get_stats && get "^${pxname},${svname}," ${stat}
+  get "^${pxname},${svname}," ${stat}
 fi
+write_log
```
see earlier comments regarding `write_log`

mprenditore committed:
debug and cache_gen changes

- Changed `debug` implementation as adviced by @anapsix
- Removed `write_loc`
- Implemented back the `floc` usage to the new caching generation func

mprenditore committed:
changed configuration management

mprenditore committed:
fixed var name

mprenditore committed:
added flock to `get` call / indent fix

mprenditore committed:
updated debug

mprenditore commented:
All changes should have been done.
Quick recap:

- changed configuration to use ENV vars
- used back the flock
- debug optimizations

mprenditore commented:
going to close this since I've much more commit to push on my master branch.
